torch/_inductor/codegen/cpp_gemm_template.py

```diff
@@ -365,33 +377,62 @@ def get_cache_blocking(register_blocking, thread_blocking):
 
             num_byte_A = get_num_byte(self.input_nodes[0].get_dtype())
             num_byte_B = get_num_byte(self.input_nodes[1].get_dtype())
 
-            size_cache_B = K0 * Kc_blocks * N0 * Nc_blocks * num_byte_B
-
-            if size_cache_B > B_size_limit:
-                Kc_blocks = math.floor(
-                    B_size_limit / (K0 * N0 * Nc_blocks * num_byte_B)
-                )
-
-            size_cache_A = M0 * Mc_blocks * K0 * Kc_blocks * num_byte_A
-            if size_cache_A > A_size_limit:
-                Mc_blocks = math.floor(
-                    A_size_limit / (M0 * Kc_blocks * K0 * num_byte_A)
-                )
+            # NOTE [CPP GEMM Cache Blocking Algorithm]
+            # Our overall strategy is to
+            # 1) Make cache blocks of B L1-reside and reused by multiple rows of A, i.e. Mc.
+            #    Here, B is Kc x Nr where Nr is a single register block. We use L1 size to
+            #    decide Kc. We want to make Mc large enough to better reuse B.
+            # 2) Make cache blocks of A L2-reside, which would limit Mc. We want to reuse A
+            #    along N, where we have two sub-strategies (see notes below) to decide Mc and Nc.
+
+            # Step 1: Decide Kc assuming B block is L1-reside.
+            size_cache_B = K0 * Kt_blocks * N0 * num_byte_B
+            Kc_blocks = Kt_blocks
+            if size_cache_B > L1:
+                Kc_blocks = math.floor(L1 / (K0 * N0 * num_byte_B))
+
+            # Step 2: Decide Mc assuming A block is L2-reside.
+            min_Mc_ratio = 2  # TODO(jgong5): something to tune?
+            min_Mc_blocks = math.ceil(min_Mc_ratio * M0 / N0)
+            assert min_Mc_blocks >= 1
+            Kt_bytes = Kt_blocks * K0 * num_byte_A
+            if min_Mc_blocks * M0 * Kt_bytes < L2:
+                # Strategy 1: A (Mc x Kt) resides in L2 and reused by all Nt
+                # when Nc_blocks is kept 1. Mc should be large enough (>= min_Mc_blocks)
+                # to reuse B (Kc x Nr) in L1. This makes C (Mc x Nr) small enough to reside
+                # in L1.
+                Mc_blocks = min(Mt_blocks, math.floor(L2 / (M0 * Kt_bytes)))
+                Nc_blocks = 1
+            else:
+                # Strategy 2: Kt is too large to hold A (Mc x Kt) in L2, we reuse
+                # A (Mc x Kc) in L2 by B (Kc x Nc). C (Mc x Nc) resides in L2.
+                Mc_blocks = Mt_blocks
+                Nc_blocks = min(math.ceil(Mc_blocks * M0 / N0), Nt_blocks)
+                Nc_bytes = Nc_blocks * N0 * 4  # assume C or acc is float32/int32
+                Kc_bytes = Kc_blocks * K0 * num_byte_A
+                if Mc_blocks * M0 * (Kc_bytes + Nc_bytes) > L2:
+                    # The following is the solution for 4*Mc*Nc + Mc*Kc_bytes = L2,
+                    # assuming Mc == Nc for good data reuse.
+                    M_max = (math.sqrt(Kc_bytes * Kc_bytes + 16 * L2) - Kc_bytes) / 8
+                    if M_max < Mc_blocks * M0:
```

Comment:
Should it be an `assert` theoretically? But considering we use some approximated calculation. I guess it should be fine.

Reply:
It is not an assertion. When it is false, we use the default one which is `Mt_blocks`.